### PR TITLE
Change design of account notes in web UI

### PR DESCRIPTION
**Pull request by Gargron**
Peek 2020-07-03 21-23

- Can press `ctrl+enter` to save and unfocus
- Can press `esc` to reset and unfocus
- Saves on blur or unmount

### CODE REVIEW

**Comment by Sasha-Sorokin** [comment]
Wow! That's nice and seamless, how does it look in light theme?

**Comment by Sasha-Sorokin** [comment]
Also,

> Can press `ctrl+enter` to save and unfocus

Why not just adapt <kbd>Enter</kbd> for this, but allow newline with <kbd>Shift</kbd>+<kbd>Enter</kbd>?

> Can press `esc` to reset and unfocus

That would be unfunny to discover on random. Maybe it's best to remove it altogether? <kbd>Esc</kbd> behavior is already unclear (sometimes it closes modals, sometimes it cancels page load).

**Comment by Gargron** [comment]
> Why not just adapt Enter for this, but allow newline with Shift+Enter?

Ctrl+Enter works as save/submit in multiple other forms so it seems like what most users would expect to just work.

> > Can press `esc` to reset and unfocus
> 
> That would be unfunny to discover on random. Maybe it's best to remove it altogether?

Not sure, but it makes sense to me to have a quick reset hotkey.

**Comment by SuperSandro2000** [comment]
> Can press ctrl+enter to save and unfocus

This is the default on many other website with similar forms including Github

Not sure about the reset on <kbd>Esc</kbd> but the unfocus is intuitiv.

**Comment by Gargron** [comment]
I doubt anyone would have a problem with it but if anyone does I'll change it in the next update

**Comment by ThibG** [comment]
It does look better, but I'm not sure if it's as intuitive or usable. It'd be the first thing we have so far that you just click to edit, and there's nothing telling you that unfocusing it will save it or anything.
Also, making it sit between two parts that are entirely defined by the account owner feels a bit weird and confusing.

**Comment by Gargron** [comment]
Actually, the media descriptions used to be like this! So even though we changed them into the media editing modal, there's precedent. I've tried different combinations and I think it sits here the best!

**Comment by ThibG** [comment]
Oh, I remember the media editing thing being an issue on mobile because of that: it was difficult discovering you could actually do that. The “Click to add a note” placeholder probably is enough to address that issue though.

**Comment by ThibG** [comment]
Yeah, it does look better, but my complaints with it are:
- it's not completely obvious that you can edit the note by clicking on it (that concern is mostly addressed by the “Click to add a note” placeholder, but that placeholder won't be there when you have a note already, from, e.g. a moved blocked person, or by adding a note yourself, from the WebUI, or from a third-party client that may have a different UI for this)
- the lack of “save” button will probably confuse a lot of users at first, especially since it's the only thing that works like this in the UI right now
- the lack of feedback regarding whether it's successfully saved or not

**Comment by ThibG** [comment]
What about keeping the same style as this PR, but initiating the edit with a pencil button or something besides the “Note” header, and having the “Save” / “Cancel” button from previously? They would increase the height of the note part of the UI, but not change how the input itself would look.

**Comment by Sasha-Sorokin** [comment]
> @Gargron 
> Ctrl+Enter works as save/submit in multiple other forms so it seems like what most users would expect to just work.

Not really, no. In context of [non-retractable] *forms*, you may be required to enter multiple lines, there this would be suitable. For user notes, if we adapting such quick edit field, I think save shortcut must be simplified too. What are the odds user is going to insert multiple line breaks *on a short note*? Pretty low, probably. “And if you are exceptional one, we give you <kbd>Shift</kbd>+<kbd>Enter</kbd>” (*[sound effect](https://www.youtube.com/embed/8URukvnUYTw)*).

---

> @ThibG 
> it's not completely obvious that you can edit the note by clicking on it 

Very good point and not easy to solve, we already have a pointer, which changes to beam on hover, but maybe can make it ever more noticeable by adding a border on hover (which is massive area).

<details><summary>Like this</summary>
<p align=center>GIF demonstrating border appears when cursor hovers over textarea, when textarea is focused, border color matches its background</p>
</details>

I'll come up with other concept a bit later, I hope.

---

Regarding save / cancel buttons, I don't see any benefit of them. The notes feature is *dead simple*, as a notepad — it's only you ~~(and instance admin, but shhh)~~ who sees them, and you probably not going to put very important stuff there, so you don't really need to worry about *submitting* it. So yea, saving on unfocus works pretty well in such seamlessly integrated textarea. Remove resetting and debounce saving on type and users will know it's saving as they type, because:

---

Also,

> the lack of feedback regarding whether it's successfully saved or not

<p align=center>GIF demonstrating “Saved” indicator appears every time note is saved (along with 500 Internal Server Errors, which suggests Sasha did something wrong on server side)</p>

<details><summary>&lt;details&gt;</summary>
I took some time to rewrite the whole class component to functional one, just for fun, you know. And thought, why shouldn't I try and implement my concept? Umm, yea… Regarding 500 errors, it's my developer instance slightly broken with me jumping from one branch to another, and this actually makes a good point that I need to determine whether it saved successfully or not, don't think it's much of a problem. I may post gist with the code, but it's too much hooking, you may not like this.
</details>

*P.S. The border on GIF is React DevTools.*

**Comment by ThibG** [comment]
> What are the odds user is going to insert multiple line breaks _on a short note_? Pretty low, probably.

Probably low, yes, but not extremely so, and the user would need a way to find out that they can indeed do that.
Everywhere else, when multiple lines are possible, you simply press “enter” to achieve them. You'd basically need knowledge of *other* software to know that shift+enter would allow you adding a newline.

> Very good point and not easy to solve, we already have a pointer, which changes to beam on hover, but maybe can make it ever more noticeable by adding a border on hover (which is massive area).

Those things only work when you have a mouse pointer, not when you're on mobile.

> Regarding save / cancel buttons, I don't see any benefit of them. The notes feature is _dead simple_, as a notepad — it's only you ~(and instance admin, but shhh)~ who sees them, and you probably not going to put very important stuff there, so you don't really need to worry about _submitting_ it. So yea, saving on unfocus works pretty well in such seamlessly integrated textarea. Remove resetting and debounce saving on type and users will know it's saving as they type, because:

Yeah, the feature is simple, but you need to know how it'll be saved, and I think you'd need to be able to cancel a change too. Currently, you can do that by pressing Escape. That doesn't work on mobile.

**Comment by Sasha-Sorokin** [comment]
> @Sasha-Sorokin 
> I'll come up with other concept a bit later, I hope.

Updated [previous comment](https://github.com/tootsuite/mastodon/pull/14208#issuecomment-653879020) with another concept (see “using hint” spoiler).

File: app/javascript/mastodon/features/account/components/account_note.js [hunk 1]
```diff
@@ -3,99 +3,111 @@ import ImmutablePropTypes from 'react-immutable-proptypes';
...
 import Textarea from 'react-textarea-autosize';
+import { is } from 'immutable';
 
 const messages = defineMessages({
-  placeholder: { id: 'account_note.placeholder', defaultMessage: 'No comment provided' },
+  placeholder: { id: 'account_note.placeholder', defaultMessage: 'Click to add a note' },
 });
 
 export default @injectIntl
 class Header extends ImmutablePureComponent {
```
Comment (Sasha-Sorokin) [inline app/javascript/mastodon/features/account/components/account_note.js:14]: <kbd>Ctrl</kbd>+<kbd>C</kbd> <kbd>Ctrl</kbd>+<kbd>V</kbd>. Maybe worth fixing :eyes:

File: app/javascript/mastodon/features/account/components/account_note.js [hunk 2]
```diff
@@ -3,99 +3,111 @@ import ImmutablePropTypes from 'react-immutable-proptypes';
...
+      <div className='account__header__account-note'>
+        <strong><FormattedMessage id='account.account_note_header' defaultMessage='Note' /></strong>
 
-    let note_container = null;
-    if (isEditing) {
-      note_container = (
         <Textarea
           className='account__header__account-note__content'
-          disabled={isSubmitting}
+          disabled={(typeof this.props.value === undefined)}
```
Comment (Sasha-Sorokin) [inline app/javascript/mastodon/features/account/components/account_note.js:103]: `typeof` never returns undefined, you meant `undefined` :)

**Comment by Gargron** [comment]
- Fixed some bugs around accidentally erasing note while the web UI loads it
- Removed blur on ctrl+enter because the consequent blur event sends a 2nd save request really fast and that's not nice
- Added "Saved" inline alert from @Sasha-Sorokin's suggestion

**Review by ClearlyClaire — COMMENTED** [review]
Fully went through the code, and I haven't seen any real issue with it, but I think the component being stateful and trying to handle context changes make it significantly more difficult to understand, tbh.

Design-wise, I think it looks significantly better than the previous version, but I'm still not convinced it's as easy to use, particularly on mobile, where you don't really have a way to cancel your changes. Maybe those concerns aren't that important though :woman_shrugging:

File: app/javascript/mastodon/features/account/components/header.js [hunk 3]
```diff
@@ -323,6 +314,8 @@ class Header extends ImmutablePureComponent {
                 </div>
               )}
 
+              {account.get('id') !== me && <AccountNoteContainer account={account} />}
```
Comment (Sasha-Sorokin) [inline app/javascript/mastodon/features/account/components/header.js:317]: Wait, why did you remove ability to set note for yourself?

Comment (Gargron) [reply app/javascript/mastodon/features/account/components/header.js:317]: Because we don't load relationship for the user itself the note is never loaded, so while it was possible to set it, it would always show up as empty.

Comment (Sasha-Sorokin) [reply app/javascript/mastodon/features/account/components/header.js:317]: Is it possible to fix it somehow? Because I find note for myself useful, just like chatting with myself in messengers… you can leave some things for later. I actively used that in Discord.

**Comment by blank71** [comment]
Add keyboard shortcut is nice. But I would like you to add some buttons, like previous one. 
https://github.com/tootsuite/mastodon/pull/14148#issuecomment-651280330

Current one is not user-friendly at mobile and for mastodon beginner.
I tried both, previous one and current one. Current one quite confused me to add note, because there are no any buttons. At this point previous one is nice.

Mastodon settings has "save" button and setting is not saved automatically. This is my thought that Mastodon base design is saving manually not saving automatically. So add "save" botton is natural for me.😊
